From 09c259399447ca0d828c65946b7d938e4692d593 Mon Sep 17 00:00:00 2001
From: Colin Ian King <colin.i.king@gmail.com>
Date: Tue, 27 May 2025 15:12:58 +0100
Subject: [PATCH 178/180] readdir: add unlikely branch hint on len check

Add an unlikely hint on the error return path check in
verify_dirent_name(). This improves performance when testing with
stress-ng dentry and dirent stressors.

Benchmarking shows:
  getdent: 1.1% improvement
  dentry:  0.9% improvement

Signed-off-by: Colin Ian King <colin.i.king@gmail.com>
---
 fs/readdir.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/fs/readdir.c b/fs/readdir.c
index 7764b8638978..c501155ed99a 100644
--- a/fs/readdir.c
+++ b/fs/readdir.c
@@ -147,7 +147,7 @@ EXPORT_SYMBOL(iterate_dir);
  */
 static int verify_dirent_name(const char *name, int len)
 {
-	if (len <= 0 || len >= PATH_MAX)
+	if (unlikely(len <= 0 || len >= PATH_MAX))
 		return -EIO;
 	if (memchr(name, '/', len))
 		return -EIO;
-- 
2.47.1

